fix(ivy): TestBed.get(Compiler) throws "Error: Runtime compiler is not loaded"

mhevery opened the pull request:
This PR adds support for the `TestBed.get(Compiler)` and unblocks FW-561.

fix(ivy): TestBed.get(Compiler) throws "Error: Runtime compiler is not loaded"


BREAKING CHANGE:

The public API for `DebugNode` was accidentally too broad. This change removes
1. Public constructor. Since `DebugNode` is a way for Angular to communicate information
   on to the developer there is no reason why the developer should ever need to
   Instantiate the `DebugNode`
2. We are also removing `removeChild`, `addChild`, `insertBefore`, and `insertChildAfter`.
   All of these methods are used by Angular to constructor the correct `DebugNode` tree.
   There is no reason why the developer should ever be constructing a `DebugNode` tree
   And these methods should have never been made public. 
3. All properties have been change to `readonly` since `DebugNode` is used by Angular 
   to communicate to developer and there is no reason why these APIs should be writable.

While technically breaking change we don’t expect anyone to be effected by this change.

mhevery committed:
fix(ivy): TestBed.get(Compiler) throws "Error: Runtime compiler is not loaded"

BREAKING CHANGE:

The public API for `DebugNode` was accidentally too broad. This change removes
1. Public constructor. Since `DebugNode` is a way for Angular to communicate information
   on to the developer there is no reason why the developer should ever need to
   Instantiate the `DebugNode`
2. We are also removing `removeChild`, `addChild`, `insertBefore`, and `insertChildAfter`.
   All of these methods are used by Angular to constructor the correct `DebugNode` tree.
   There is no reason why the developer should ever be constructing a `DebugNode` tree
   And these methods should have never been made public.
3. All properties have been change to `readonly` since `DebugNode` is used by Angular
   to communicate to developer and there is no reason why these APIs should be writable.

While technically breaking change we don’t expect anyone to be effected by this change.

mary-poppins commented:
You can preview e1b3860 at https://pr27223-e1b3860.ngbuilds.io/.

pkozlowski-opensource commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -16,12 +21,26 @@ export class EventListener {
-export class DebugNode {
-  listeners: EventListener[] = [];
-  parent: DebugElement|null = null;
+export interface DebugNode {
```
This is a breaking change, technically speaking, right? I mean, previously people could import and instantiate `DebugNode` and it was part of the public API.

I don't suspect that many people actually do it but we should still document this as a breaking change

mhevery replied:
@IgorMinar 

From my point of view we have accidentally exposed more API than we intended to. 
- We exposed the constructor
- We exposed the mutation methods 

By changing this from `class` to `interface` I think we are only removing the API which was accidentally exposed and which serves no purpose to the end user.

JoostK replied:
@mhevery Could you add to the list of breaking changes that `instanceof DebugNode` checks can no longer be done.

mhevery replied:
@JoostK will add.

IgorMinar replied:
@mhevery it should be fine to change this behavior, but we need to document it.

pkozlowski-opensource commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -146,27 +180,198 @@ function _queryElementChildren(
+  }
+  get context(): any { throw notImplemented(); }
+
+  get references(): {[key: string]: any;} {
```
Couldn't we use https://github.com/angular/angular/blob/dc300c5c4172e1f8a6f3f84da6797e427e3626c0/packages/core/src/render3/discovery_utils.ts#L185-L193 here?

mhevery replied:
you are right, fixed

pkozlowski-opensource commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -146,27 +180,198 @@ function _queryElementChildren(
+    return properties;
+  }
+
+  get attributes(): {[key: string]: string | null;} { throw notImplemented(); }
```
What would be the strategy for implementing it in the future? Just get static attrs from `TView`? Would we also need info about attr bindings (`attr.foo="expr"`) here? Same applies to classes and styles.

Generally speaking I'm not clear how the current view engine works in this respect...

mhevery replied:
We have all of the information in `TNode` it is just a matter of getting it. I am not sure what the current behavior is either. sounds like we need to do some investigation.

pkozlowski-opensource commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -146,27 +180,198 @@ function _queryElementChildren(
+  }
+
+  triggerEventHandler(eventName: string, eventObj: any): void {
+    const event = document.createEvent('MouseEvent');
```
Uh, this is big change as compared to the current implementation that tries to abstract native event handlers. Not only the `'MouseEvent'` is hard-coded here but I suspect that it will fail for pseudo-events like `window.keydown.ESC`.


Also I'm not sure if the current implementation triggers event handlers on directives.

tl;dr; is that I was trying to take the approach you are taking in this PR and decided agains party for the above reasons. The existing `DebugElement` abstracts native platform while here we are going low-level native. `TestBed` is a public API and this change would be very breaking for many projects I know (ex. ng-bootstrap).

As a very minimum we need to evaluate this breaking change and document if we decide to go down this path.

mhevery replied:
sorry, this is a temporary hack implementation. Proper implementation to be added. I have added comments to document it.

mhevery replied:
https://angular-team.atlassian.net/browse/FW-719

pkozlowski-opensource commented on packages/core/src/render3/context_discovery.ts:
```diff
@@ -262,12 +262,6 @@ function findViaDirective(lViewData: LViewData, directiveInstance: {}): number {
   return -1;
 }
 
-function assertDomElement(element: any) {
```
👍 good move moving this one :-)

pkozlowski-opensource commented on packages/core/src/view/services.ts:
```diff
@@ -780,14 +780,14 @@ export class DebugRenderer2 implements Renderer2 {
     const el = this.delegate.selectRootElement(selectorOrNode, preserveContent);
     const debugCtx = getCurrentDebugContext() || (ivyEnabled ? this.createDebugContext(el) : null);
```
I believe that we can remove the part guided by `ivyEnabled` now.

mhevery replied:
Good catch.

I removed the check

pkozlowski-opensource reviewed (changes requested):
@mhevery I've left few comments in the code, the most important being about breaking changes. I was trying to take the same approach in the past but I got scared off the breaking changes with triggering event handlers etc. Basically we would loose the abstraction over native platform and this sounds like "big deal". Let's discuss.

If we decide to go ahead with this approach it is not clear for me what should happen with https://github.com/angular/angular/blob/master/packages/core/src/render3/debug.ts and all infrastructure that plugs it (https://github.com/angular/angular/blob/master/packages/core/testing/src/r3_test_bed.ts#L426-L429)  

Lastly, the CI is red with some tests failing.

mary-poppins commented:
You can preview afe9b25 at https://pr27223-afe9b25.ngbuilds.io/.

mhevery commented:
> If we decide to go ahead with this approach it is not clear for me what should happen with https://github.com/angular/angular/blob/master/packages/core/src/render3/debug.ts and all infrastructure that plugs it (https://github.com/angular/angular/blob/master/packages/core/testing/src/r3_test_bed.ts#L426-L429)

Ahh that is an excellent Catch! Yes that code is dead and I have now deleted it. :-)

mary-poppins commented:
You can preview bb8f9cb at https://pr27223-bb8f9cb.ngbuilds.io/.

pkozlowski-opensource reviewed (approved):
@mhevery  if you are positive that we can implement the rest of the `DebugElement` methods this LGTM
Would be probably good if someone else can do another pass on it as we are touching many parts

mhevery commented:
http://test/OCL:222839683:BASE:222839704:1543254316995:47c83647

kara commented on packages/common/test/directives/ng_component_outlet_spec.ts:
```diff
@@ -144,22 +144,21 @@ describe('insert/remove', () => {
+     }));
 
-  fixmeIvy('FW-561: Runtime compiler is not loaded') &&
+  fixmeIvy('destroy is not being called') &&
```
Can you add the Jira ticket number associated with this bug or mark it "unknown"? Otherwise it's hard to tell whether this root cause has been triaged.

mhevery replied:
destroy on NgModuleRef is not being called

kara commented on packages/core/src/render3/assert.ts:
```diff
@@ -83,3 +83,7 @@ function throwError(msg: string): never {
 }
+
+export function assertDomNode(node: any) {
+  assertEqual(node instanceof Node, true, 'The provided value must be an instance of an DOM Node');
```
```suggestion
  assertEqual(node instanceof Node, true, 'The provided value must be an instance of a DOM Node');
```

mhevery replied:
fixed

kara commented on packages/core/test/linker/ng_module_integration_spec.ts:
```diff
@@ -284,39 +276,36 @@ function declareTests(config?: {useJit: boolean}) {
-            expect(factory).toBeTruthy();
-            expect(factory.moduleType).toBe(SomeModule);
-          });
+      fixmeIvy('No module with ID myid loaded') && it('should register loaded modules', () => {
```
Reason should have Jira ticket or "unknown" here?

mhevery replied:
https://angular-team.atlassian.net/browse/FW-740

mhevery committed:
build: ts-api-guardian should support interface with value types

This fixes an issue where a value would hide the type.

```
export interface Foo {
  someMethod(): void;
}

export const Foo: Function = ...;
```

In the above example the `Foo` constant will hide the `interface Foo` symbol.
This change properly saves the interface in addition to the type.

mary-poppins commented:
You can preview c479d06 at https://pr27223-c479d06.ngbuilds.io/.

mary-poppins commented:
You can preview dfbf641 at https://pr27223-dfbf641.ngbuilds.io/.

mary-poppins commented:
You can preview 2564f80 at https://pr27223-2564f80.ngbuilds.io/.

mary-poppins commented:
You can preview 07c95ec at https://pr27223-07c95ec.ngbuilds.io/.

IgorMinar commented on packages/core/src/metadata/directives.ts:
```diff
@@ -77,6 +77,9 @@ export interface DirectiveDecorator {
 }
 
+/**
+ * @publicApi
```
why do we declare and export directive interface twice? that looks really fishy. Can we remove one?

IgorMinar commented on tools/public_api_guard/core/testing.d.ts:
```diff
@@ -47,6 +47,52 @@ export declare type MetadataOverride<T> = {
 
 export declare function resetFakeAsyncZone(): void;
 
+export interface TestBed {
```
I can't believe that we missed having all of these interfaces in the guard. Good catch!!!

mhevery replied:
fixed

IgorMinar reviewed (approved):
The rest looks good

mary-poppins commented:
You can preview 8ecc627 at https://pr27223-8ecc627.ngbuilds.io/.

mhevery commented:
http://test/OCL:222881368:BASE:222938072:1543299184520:907394cc

mary-poppins commented:
You can preview 9028161 at https://pr27223-9028161.ngbuilds.io/.

pkozlowski-opensource commented:
@mhevery looks like CI is red (the golden symbols file need updating etc.)

mhevery commented:
http://test/OCL:222839683:BASE:223003683:1543339823046:c3644f5e

mary-poppins commented:
You can preview 8ee99af at https://pr27223-8ee99af.ngbuilds.io/.

mary-poppins commented:
You can preview 7e65c8a at https://pr27223-7e65c8a.ngbuilds.io/.

mary-poppins commented:
You can preview 8504136 at https://pr27223-8504136.ngbuilds.io/.

mhevery committed:
fixup! fix(ivy): TestBed.get(Compiler) throws "Error: Runtime compiler is not loaded"

mary-poppins commented:
You can preview f6f9612 at https://pr27223-f6f9612.ngbuilds.io/.
